scalameta/trees/shared/src/main/scala/scala/meta/Trees.scala

```diff
@@ -330,9 +330,11 @@ object Ctor {
   checkParent(ParentChecks.Init)
 }
 
+@ast class Self(name: Term.Name, decltpe: Option[Type]) extends Member
+
 @ast class Template(early: List[Stat],
                     inits: List[Init],
-                    self: Term.Param,
+                    self: Self,
```

Comment:
Interestingly, the old design didn't guard against non-sensical trees like 

```scala
@ Template(Nil, Nil, Term.Param(List(mod"lazy"), Term.Name("b"), Some(Type.Name("X")), Some(q"2")), Some(Nil))
res7: Template = { lazy b: X = 2 => }
```

This is a nice improvement.